### DISCUSSION TITLE
Language selection?

### DISCUSSION THREAD

**Discussion by alcohen** [VolksEEG/VolksEEG#10] [Firmware]
Arduino, C or Rust?

It sounds pretty attractive to me to _start_ with a Rust "framework" and then do the detail work in C, then start moving from C->Rust as we can. I guess my questions would be:

1. How much time/effort would this take compared to doing in just Arduino or C? Or just doing it in Rust (including overcoming learning curve)? Is it possibly a matter of Kevin writing the "framework" then the rest of us fill in with C or Rust as we can?

2. Would adding Rust make this more interesting/fun for developers?

Generally speaking, for this prototype, speed is important thing. But not the only thing.

**Comment by stevebranam** [comment]
@alcohen and I were talking briefly after the meeting about this. I agree that the hybrid approach seems like a good combination to avoid too much delay due to learning curve, yet start on the path of using Rust. That does seem like a safety-oriented approach to take the next step for the industry.

@Nauscar has provided a good starting framework. If he adds sufficient "call C code here" hooks, outlining whatever constraints are required to use that, I think we could use it fairly quickly.

As a practical matter, now holidays are coming up, and according to Hofstadter's Law, "It always takes longer than you expect, even when you take into account Hofstadter's Law." So I suggested I can take 2-3 weeks to coordinate with Kevin, go through some Rust resources, and do some proof-of-concept of the prototype.

We know this application is a fairly simple data acquisition system. I have no experience with that type of system, but I'm familiar with the "data pump" architectural model from network routers. That's like a 2-stroke engine: intake stroke, acquire data (or a data frame across all channels); output stroke, push the data out to the consumer, in this case the PC USB connection. That can be pipelined with suitable elastic buffering for rate matching.

At the moment I have 2 books available that I'll skim today and tomorrow (my blog post on skimming: [Using Speed Reading Skimming](https://www.vivocentum.com/2021/10/using-speed-reading-skimming.html)):

1. The rust-lang.org online book: [The Rust Programming Language](https://doc.rust-lang.org/book/title-page.html)
2. Hardcopy: [Beginning Rust: From Novice to Professional](https://www.amazon.com/Beginning-Rust-Professional-Carlo-Milanesi-dp-1484234693/dp/1484234693/)

**Comment by stevebranam** [comment]
I found a great starter video on Rust. Here's [my mini-review](https://www.linkedin.com/posts/steve-branam-3112319_rust-crash-course-rustlang-activity-6863452817424732160-weXO). I watched it at 2x during my morning workout and breakfast, very easy to follow for any experienced developer.

This is totally doable! We'll just need to have @Nauscar help us with the specifics of the embedded system environment that he demoed to us, big thank you to Kevin!

Also, I note that the guy doing the video is fairly local to me and @alcohen, in Amesbury, MA. Sounds like a coffee trip!

**Reply by alcohen** [reply]
Thanks Steve! I'll check it out.

**Comment by longevity-software** [comment]
1: Having had some time to think about it, I am not sure it would take much longer, in my head, I compare it to TDD, you may spend more time upfront, but you will save debugging time going forward. 
I think a hybrid approach sounds good, but may complicate things, depending how the C bits are written it may take just as long again to rewrite. 

2: yes. 

In my opinion anyway, it's a language I would like to learn so it would definitely be more interesting for me. 

I can see the benefits of hybrid approach, allowing more people to contribute to the development of the device, we need to be diligent about bringing everything to rust as soon as we can though, otherwise we will just end up with a rusty-C soup.

**Reply by stevebranam** [reply]
After watching the video above, I think there's enough similarity to C/C++/Python that it's not that difficult a stretch to use Rust fully, not hybrid. While @Nauscar's demo had some odd-looking syntax, once we get past that, it looks pretty straightforward.

We'll inevitably have a few stumbles as we shift our thinking a bit, but shouldn't be that difficult for experienced developers to adjust.

**Reply by alcohen** [reply]
I'm looking forward to seeing how the hybrid works in practice in day-to-day usage; is it simply a matter of writing C code to address well-defined interfaces? 

The thing that I'm most thoughtful of with regard to using only Rust is that we could get into situations, particularly when dealing with hardware, where it's not clear how to do things in Rust but we know how to do it in C. 1-800-ASK-KEVIN might sufficiently mitigate that.

And I basically agree on comparison with TDD.

**Comment by stevebranam** [comment]
@Nauscar do you have any recommendations for books/videos/blogs/tutorials on Rust specific to embedded systems?

**Reply by Nauscar** [reply]
A list of awesome embedded rust books, blogs and trainings: https://github.com/rust-embedded/awesome-embedded-rust#books-blogs-and-training-materials

The user documentation for RTIC is quite extensive (I found the tips & tricks section really interesting): https://rtic.rs/dev/book/en/

**Reply by stevebranam** [reply]
@Nauscar That looks like an awesome collection, lots of information! From all that, is there a directed path you would choose to navigate through it? Given that we're all starting from scratch on Rust, can you curate a roadmap for a fast, focused ramp-up? And if there are a few side paths worth keeping in mind for later, indicate those? As well as some stuff to ignore for now.

If you have the time to get specific, you could even stitch together a reading list:

1. Start here, these main sections
2. Then here, these sections (skip xyz for now)
3. Then this PDF
4. etc, etc.

Even just parceling it out 2 or 3 items at a time. Eventually that turns into your full mapped-out list. I know it's time consuming to do all that list curation and creation from doing it for my blog series [Video-Based STEM Embedded Systems Curriculum](https://www.embeddedrelated.com/showarticle/1425.php)

Your list could later turn into a similar curriculum blog post for embedded Rust development. As always, embedded development is a huge field, so the key is picking a narrow focus for the path, which gives people the foundation for exploring further on their own.

**Comment by stevebranam** [comment]
Some Rust resources recommended by one of my coworkers:

This podcast is really, really good for learning to grok Rust (not specifically for embedded systems): https://newrustacean.com
It mostly covers core features, so even though it ended a while ago, it's still a great primer.

https://ferrous-systems.com/ These folks are THE Embedded Rust folks, and if you don't already follow James Munns (one of their principals), he is practically THE Embedded Rust guy: https://twitter.com/bitshiftmask

https://oxidizeconf.com is the first embedded rust conference that I know of, I attended their event in 2020, and it was fantastic. Highly recommend looking up some of their talks on YouTube.

The embedded rustaceans largely organize themselves on Matrix, with other contact info here: https://github.com/rust-embedded/wg, and are all superb at answering questions. Smaller embedded rust communities also tend to live on Matrix (there are communities for the STM32 and the SAM-series Cortex-M processors, that I know of).

ferrous-systems.comferrous-systems.com
Ferrous Systems
Ferrous Systems is a Berlin based Rust Consultancy

twitter.comtwitter.com
PhantomDataᐸJamesᐳ (@bitshiftmask) | Twitter
The latest Tweets from PhantomDataᐸJamesᐳ (@bitshiftmask). Chief Entropy Generator and Strategic Procrastinator at @FerrousSystems He/Him. Kein Ort für Nazis. Berlin, Germany

rust-embedded/wg

**Comment by stevebranam** [comment]
This is where I got started thinking about Rust, a webinar by Feabhas a couple years ago. Their slides:
[Rust-post-webinar.pdf](https://github.com/VolksEEG/VolksEEG/files/7501613/Rust-post-webinar.pdf)

And some notes from their CEO in response to my questions:

Regarding a reference for C2Rust, the most useful I've found is:
https://locka99.gitbooks.io/a-guide-to-porting-c-to-rust/content/

It's also worth looking at this presentation as it gives some useful insights:
https://github.com/carols10cents/rust-out-your-c-talk

**Comment by stevebranam** [comment]
I'm halfway through skimming "Beginning Rust". It's good, basically like a verbose version of K&R, so an excellent follow up to the video I linked to in my LinkedIn post above. Overall, definitely feels like anyone with C/C++/Python experience could be reasonably proficient quickly.

**Comment by stevebranam** [comment]
Here's another video recommendation, an excellent follow up to yesterday's. I watched about half of the videos on this playlist at 2x this morning during my workout and breakfast. He's a faster talker, so you have to focus more carefully. https://www.linkedin.com/posts/steve-branam-3112319_rust-tutorial-youtube-activity-6863817759340470273-gfuO

**Comment by stevebranam** [comment]
I'm really lovin' Rust the more I learn about it. I finished skimming _Beginning Rust_ and am now in the middle of _The Rust Programming Language_, then I'll skim _The Embedded Rust Book_. Part of why I like it so much is that I've proven empirically that all possible types of bugs **will** eventually occur.

This is a nice short video on using Test Driven Development with Rust, watchable at 2x: [Rust Testing and TDD - An Intro to Testing and Test Driven Development](https://www.youtube.com/watch?v=2vBQFIWl36k). That's why we want a modular architecture, that helps make it testable. While this video doesn't specifically mention, it looks like you can implement fakes of various kinds inside the test module, and pass them in for use by the Code Under Test via dependency injection.

**Reply by alcohen** [reply]
"I've proven empirically that all possible types of bugs will eventually occur"

I've been able to replicate your work.

**Reply by stevebranam** [reply]
Excellent, that means we're not fodder for the _Journal Of Irreproducible Results_!

**Comment by stevebranam** [comment]
I'm becoming more and more convinced that Rust is absolutely the way we want to go. This is a spectacularly magnificent introduction to embedded Rust from Frans Skarman. Thank you @Nauscar for recommending the Ferrous Systems resources!
[An Overview of the Embedded Rust Ecosystem](https://www.youtube.com/watch?v=vLYit_HHPaY), watchable to 1.5 or 2x.

**Comment by stevebranam** [comment]
Further following up on embedded Rust, this is a very interesting discussion on using it in safety critical systems. Interesting perspective from a Green Hills guy "never want to program in C++ again!" A little long, but watchable at 1.5-2x.
[Rust in Safety Critical Systems Panel](https://www.youtube.com/watch?v=g7VVM_t72NE)

**Comment by stevebranam** [comment]
My blog post on it: [Learning Rust For Embedded Systems](https://www.embeddedrelated.com/showarticle/1432.php)

Thank you, @Nauscar!

**Reply by alcohen** [reply]
> My blog post on it: [Learning Rust For Embedded Systems](https://www.embeddedrelated.com/showarticle/1432.php)
> 
> 
> 
> Thank you, @Nauscar!

Another really good column Steve￼. I love how these are logging our technical journey.￼ Do you have a Twitter account? I'd like to post the link on Twitter and reference you.￼

**Reply by stevebranam** [reply]
Thank you! I deleted my Facebook and Twitter accounts last year as being corrosive to society (and you thought GOTO was considered harmful!). You can reference my LinkedIn account https://www.linkedin.com/in/steve-branam-3112319/

**Reply by alcohen** [reply]
I don't blame you! I find that Twitter is only good as write only.￼

**Comment by jsosulski** [comment]
Hi! For what it's worth, I read about your project from the Rust mailing list, so that would be a yes for 2. for me! (Also if you consider this for the PC software)

Maybe offtopic in this discussion, but I work with EEG-based BCI, and often people use the [LabStreamingLayer](https://github.com/sccn/labstreaminglayer) to synchronize EEG and other streams for the experiments. If your device would be able to send data to LSL from the start, 1) researchers could use your device directly in their experiments that are built on LSL and 2) you could use the available LSL resources right away, see e.g.:
- [Viewer software](https://labstreaminglayer.readthedocs.io/info/viewers.html)
- [Data recorder](https://github.com/labstreaminglayer/App-LabRecorder)
- [Reference: OpenBCI LSL support](https://docs.openbci.com/Software/CompatibleThirdPartySoftware/LSL/)

In the same vein, if you want to write your own EEG-software from scratch and you build a nice one that works with LSL, other people with other EEG devices could use your visualization software right away.

Note however that I am not experienced in hardware design / dont know if LSL would create some excessive overhead, impact on performance or anything.

I did not find a mention of LSL in this repo, so my apologies if you already discussed this somewhere and I did not find it. I am also not affiliated with LSL, just interested in Rust and BCIs :)

**Reply by alcohen** [reply]
Hi, thanks for coming by. LSL sounds interesting, I'm going to take a look now. Do you have a specific interest in VolksEEG, either in contributing or using it?￼

**Reply by jsosulski** [reply]
LSL often advertises itself with "streaming data over the network", but in our research group we used it exclusively as a way to unify handling of streamed data on the same local machine, e.g., eye tracker and EEG signal.

I found VolksEEG via the "This Week in Rust" Mailing list and stevebranam's blog post on there. I would be interested in contributing, but my experience in embedded development is slim and in embedded Rust it is virtually nonexistent. However, I plan to get an NRF5340 and familiarize myself with Rust on that MCU, maybe I can be more helpful in the future.

**Reply by stevebranam** [reply]
Hi Jan, If you'd like to learn some embedded system stuff and doing it the Rust way, the Discovery Book looks like a very good practical, hands-on introduction. It uses the STM32F303 Discovery Kit, which I got from Mouser for $16 plus $8 shipping. [https://docs.rust-embedded.org/discovery/index.html](https://docs.rust-embedded.org/discovery/index.html)

**Reply by stevebranam** [reply]
Also, The Embedded Rust book covers using QEMU to emulate real HW on your laptop, so you can try out the embedded stuff without the board. https://docs.rust-embedded.org/book/intro/index.html

**Reply by jsosulski** [reply]
The discovery book looks interesting, thank you! But I am not a fan of emulation, when I do something on hardware, I want to have the bare metal in front of me. Also there is no possibility (I hope at least) to create some [magic smoke](https://en.wikipedia.org/wiki/Magic_smoke) with emulators :-)
